lib/crawly/request.ex

```diff
@@ -2,22 +2,184 @@ defmodule Crawly.Request do
   @moduledoc """
   Request wrapper
 
-  Defines Crawly request structure.
+  Defines Crawly request structure, and API to access Crawly.Request fields
   """
-  defstruct url: nil, headers: [], prev_response: nil, options: []
+
+  ###===========================================================================
+  ### Type definitions
+  ###===========================================================================
+  defstruct url: nil,
+            headers: [],
+            prev_response: nil,
+            options: [],
+            retries: 0,
+            middlewares: []
 
   @type header() :: {key(), value()}
   @type url() :: binary()
 
   @typep key :: binary()
   @typep value :: binary()
-
+  @type retries :: non_neg_integer()
   @type option :: {atom(), binary()}
+  @opaque t :: %__MODULE__{
+                 url: binary(),
+                 headers: [header()],
+                 prev_response: %{},
+                 retries: 0,
+                 middlewares: [],
+                 options: [option()]
+               }
+  ###===========================================================================
+  ### API functions
+  ###===========================================================================
+  @doc """
+  Create new Crawly.Request from url, headers and options
+  """
+  @spec new(url, headers, options) :: request
+        when url: binary(),
+             headers: [term()],
+             options: [term()],
+             request: Crawly.Request.t()
+
+  def new(url, headers \\ [], options \\ []) do
+    # Define a list of middlewares which are used by default to process
+    # incoming requests
+    default_middlewares = [
+      Crawly.Middlewares.DomainFilter,
+      Crawly.Middlewares.UniqueRequest,
+      Crawly.Middlewares.RobotsTxt
+    ]
+
+    middlewares =
+      Application.get_env(:crawly, :middlewares, default_middlewares)
+
+    new(url, headers, options, middlewares)
+  end
+
+  @doc """
+  Same as Crawly.Request.new/3 from but allows to specify middlewares as the 4th
+  parameter.
+  """
+  @spec new(url, headers, options, middlewares) :: request
+        when url: binary(),
+             headers: [term()],
+             options: [term()],
+             middlewares: [term()], # TODO: improve typespec here
+             request: Crawly.Request.t()
+  def new(url, headers, options, middlewares) do
+    %Crawly.Request{
+      url: url,
+      headers: headers,
+      options: options,
+      middlewares: middlewares
+    }
+  end
+
+  @doc """
+  Access url field from Crawly.Request
+  """
+  @spec url(request) :: url_field
+        when request: Crawly.Request.t(),
+             url_field: binary()
+  def url(request), do: request.url
+
+  @doc """
+  Set url field in Crawly.Request
+  """
+  @spec url(request, new_url) :: url_field
```

Comment:
This function seems unnecessary, the request struct can be adjusted by the user directly.

Reply:
@Ziinc lets address all concerns to API based requests altogether in the main idea explanation comment